torch/optim/optimizer.py

```diff
@@ -164,18 +169,82 @@ def zero_grad(self):
                     p.grad.detach_()
                     p.grad.zero_()
 
-    def step(self, closure):
-        r"""Performs a single optimization step (parameter update).
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step (parameter update).
 
         Arguments:
-            closure (callable): A closure that reevaluates the model and
-                returns the loss. Optional for most optimizers.
+            closure (callable, optional): A closure that reevaluates the model
+            and returns the loss. Optional for most optimizers.
+
+        .. note::
+            Unless otherwise specified, this function should not modify the
+            parameter or its ``.grad`` field.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            lr = group['lr']
+            weight_decay = group['weight_decay']
+            for p in group['params']:
+                grad = p.grad
+                if grad is None:
+                    continue
+
+                if grad.is_sparse:
+                    if weight_decay > 0:
+                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
+
+                    update = self.get_sparse_update(p, **group)
+                else:
+                    if weight_decay > 0:
+                        grad = grad.add_(weight_decay, p)
```

Comment:
Replaced by `grad = grad.add(...)`.

Reply:
the in-place addition is/was intentional, since I tried to get rid of the `grad` parameter in `get_update`, so there was no other way to use the updated gradient within `get_update`. 

The other reason was that this in-place operation would behave as if the L2 regularisation is implemented as the additional loss term it actually is.